share/eds/store.go

```diff
@@ -255,17 +258,35 @@ func (s *Store) put(ctx context.Context, root share.DataHash, square *rsmt2d.Ext
 		return fmt.Errorf("failed to initiate shard registration: %w", err)
 	}
 
+	var result dagstore.ShardResult
 	select {
+	case result = <-ch:
 	case <-ctx.Done():
 		// if context finished before result was received, track result in separate goroutine
 		go trackLateResult("put", ch, s.metrics, time.Minute*5)
 		return ctx.Err()
-	case result := <-ch:
-		if result.Error != nil {
-			return fmt.Errorf("failed to register shard: %w", result.Error)
-		}
-		return nil
 	}
+
+	if result.Error != nil {
+		return fmt.Errorf("failed to register shard: %w", result.Error)
+	}
+
+	// accessor returned in result will be nil, so shard needs to be acquired first, to become
+	// available in cache. It might take some time and result should not affect put operation, so do it
+	// in goroutine
+	//TODO: Ideally only recent blocks should be put in cache, but there is no way right now to check
```

Comment:
after pruning, only blocks inside the recency window will be put anyways. So I agree that this is not a big issue. It only happens during initial sync anyways